### ISSUE TITLE
assertEqualsDeep(Set, Set, String) cut short on checking actual set's size

### ISSUE THREAD

**Issue by edrik** [cbeust/testng#1917]
### TestNG Version

Latest version

### Expected behavior
For:
```
assertEqualsDeep( new HashSet<>( Arrays.asList("foo", "bar", "baz") ), new HashSet<>( Arrays.asList("foo", "bar") ), "same same" );
assertEqualsDeep( new HashSet<>( Arrays.asList("foo", "bar") ), new HashSet<>( Arrays.asList("foo", "bar", "baz") ), "same same" );
```
Expecting the first assertion to fail with:
`java.lang.AssertionError: Sets do not have the same size:2 != 3`
or
`java.lang.AssertionError: Sets not equal: expected: [bar, foo] and actual: [bar, foo, baz]`

### Actual behavior

Only the second line fails with:
`java.lang.AssertionError: Sets not equal: expected: [bar, foo, baz] and actual: [bar, foo]`

### Is the issue reproductible on runner?

- [ ] Shell
- [ ] Maven
- [x] Gradle
- [ ] Ant
- [x] Eclipse
- [ ] IntelliJ
- [ ] NetBeans

### Test case sample
```
assertEqualsDeep( new HashSet<>( Arrays.asList("foo", "bar", "baz") ), new HashSet<>( Arrays.asList("foo", "bar") ), "same same" );
assertEqualsDeep( new HashSet<>( Arrays.asList("foo", "bar") ), new HashSet<>( Arrays.asList("foo", "bar", "baz") ), "same same" );
```

**Comment by reachbharathan** [comment]
@edrik 
In the first assertion the expected set is "foo", "bar" , Actual set is "foo", "bar", "baz".
It is validating the whether "foo" and "bar" are present inside the actual set which has "foo", "bar", "baz". As "foo" and "bar" are present inside actual there is no assertion error.

Whereas in the second case the expected set is "foo", "bar" and "baz", the Actual set is "foo", "bar". On validating it is able to find "foo" and "bar", while "baz" is not present in the actual. Hence the validation error.

if we want the comparison of the exact object match then we can use assertEquals();

**Comment by edrik** [comment]
I was comparing the behaviour with assertEqualsDeep() for Map's, where the order of expected and actual value has the same outcome. And i couldn't find any documentation where it stated that the order is significant in this case.

**Comment by reachbharathan** [comment]
@edrik The issue is with the documentation, Can i we go ahead and update the documentation that closes this issue? @krmahadevan your thoughts?

**Comment by krmahadevan** [comment]
@reachbharathan doesn't your PR https://github.com/cbeust/testng/pull/1943 fix this problem? Yes we can always enrich the documentation (javadocs) with a PR.

**Comment by krmahadevan** [comment]
@reachbharathan - I dont think i quite agree with this.

`deepEquals()` in my opinion should work the same way in either cases i.e.,

if expected.size > actual.size (or)
if expected.size < actual.size

It should always first check for the size before it commences evaluation of equality of element. 

> The deepEquals is similar to containsString assert.

What do you mean here ? I didn't quite get that.

@juherr your thoughts on this ?

**Comment by juherr** [comment]
`assertEquals` check if objects are equal. 2 collections with a different size are not equal.

If we need more flexible assertions, we recommend to use AssertJ which has such kind of assertions.

**Comment by jiong-fu** [comment]
@krmahadevan @juherr it seems that this issue has been resolved with the pull request #1943 merged in December 2018, or anything remains to be implemented here?

**Comment by juherr** [comment]
Closed by #1943

**Comment by juherr** [comment]
Thank @jiong-fu 👍
